File: app/src/main/java/de/westnordost/streetcomplete/data/osm/tql/TagFilterExpression.java
package de.westnordost.streetcomplete.data.osm.tql;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import de.westnordost.osmapi.map.data.BoundingBox;
import de.westnordost.osmapi.map.data.Element;

/** Represents a parse result of a string in filter syntax, i.e.
 *  <tt>"ways with (highway = residential or highway = tertiary) and !name"</tt> */
public class TagFilterExpression
{
	private final List<ElementsTypeFilter> elementsTypeFilters;
	private final BooleanExpression<OQLExpressionValue> tagExprRoot;

	public TagFilterExpression(List<ElementsTypeFilter> elementsTypeFilters,
							   BooleanExpression<OQLExpressionValue> tagExprRoot)
	{
		this.elementsTypeFilters = elementsTypeFilters;
		this.tagExprRoot = tagExprRoot;
	}

	/** @return whether the given element is found through (=matches) this expression */
	public boolean matches(Element element)
	{
		Element.Type eleType = element.getType();

		boolean elementTypeMatches = false;

		switch(eleType)
		{
			case NODE:
				elementTypeMatches = elementsTypeFilters.contains(ElementsTypeFilter.NODES);
				break;
			case WAY:
				elementTypeMatches = elementsTypeFilters.contains(ElementsTypeFilter.WAYS);
				break;
			case RELATION:
				elementTypeMatches = elementsTypeFilters.contains(ElementsTypeFilter.RELATIONS);
				break;
		}

		return elementTypeMatches && tagExprRoot.matches(element);
	}

	/** @return this expression as a Overpass query string (in a short one-liner form) */
	public String toOverpassQLString(BoundingBox bbox)
	{
		StringBuilder oql = new StringBuilder();
		if(bbox != null)
		{
			oql.append(OverpassQLUtil.getGlobalOverpassBBox(bbox));
		}

		BooleanExpression<OQLExpressionValue> expandedExpression = createExpandedExpression();

		List<String> elements = getTagFiltersOverpassList(expandedExpression);

		final boolean useUnion = elementsTypeFilters.size() > 1 || elements.size() > 1;

		if(useUnion) oql.append("(");
		for(ElementsTypeFilter filter : elementsTypeFilters)
		{
			oql.append(getTagFiltersOverpassString(filter, elements));
		}
		if(useUnion) oql.append(");");

		return oql.toString();
	}

	private BooleanExpression<OQLExpressionValue> createExpandedExpression()
	{
		BooleanExpression<OQLExpressionValue> result = tagExprRoot.copy();
		result.flatten();
		result.expand();
		return result;
	}

	private static String getTagFiltersOverpassString(ElementsTypeFilter elementType, List<String> elements)
	{
		StringBuilder oql = new StringBuilder();
		for(String element : elements)
		{
			oql.append(elementType.oqlName);
			oql.append(element);
		}
		return oql.toString();
	}

	private static List<String> getTagFiltersOverpassList(
			BooleanExpression<OQLExpressionValue> expandedExpression)
	{
		BooleanExpression<OQLExpressionValue> child = expandedExpression.getFirstChild();
		if(child == null) return Collections.singletonList(";");

		if(child.isOr())
		{
			return getUnionOverpassQueryString(child);
		}
		else if(child.isAnd() || child.isValue())
		{
			return Collections.singletonList(getSingleOverpassQueryString(child));
		}

		throw new RuntimeException("The boolean expression is not in the expected format");
	}

	private static List<String> getUnionOverpassQueryString(BooleanExpression<OQLExpressionValue> child)
	{
		List<String> result = new ArrayList<>();
		for(BooleanExpression<OQLExpressionValue> orChild : child.getChildren())
		{
			result.add(getSingleOverpassQueryString(orChild));
		}
		return result;
	}

	private static String getSingleOverpassQueryString(BooleanExpression<OQLExpressionValue> child)
	{
		if(child.isValue()) return child.getValue().toOverpassQLString() + ";";

		if(!child.isAnd())
			throw new RuntimeException("The boolean expression is not in the expected format");

		StringBuilder result = new StringBuilder();
		for(BooleanExpression<OQLExpressionValue> valueChild : child.getChildren())
		{
			result.append(valueChild.getValue().toOverpassQLString());
		}
		result.append(";");

		return result.toString();
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/tql/TagFilterExpression.kt
package de.westnordost.streetcomplete.data.osm.tql

import de.westnordost.osmapi.map.data.Element
import de.westnordost.streetcomplete.data.osm.tql.ElementsTypeFilter.NODES
import de.westnordost.streetcomplete.data.osm.tql.ElementsTypeFilter.WAYS
import de.westnordost.streetcomplete.data.osm.tql.ElementsTypeFilter.RELATIONS

/** Represents a parse result of a string in filter syntax, i.e.
 *  "ways with (highway = residential or highway = tertiary) and !name"  */
class TagFilterExpression(
    private val elementsTypes: List<ElementsTypeFilter>,
    private val tagExprRoot: BooleanExpression<TagFilter, Tags>?
) {
    private val overpassQuery: String

    init {
        val isNwr = elementsTypes.size == 3 && elementsTypes.containsAll(ElementsTypeFilter.values().asList())
        val oqlElementNames = if (isNwr) listOf("nwr") else elementsTypes.map { it.oqlName }
        overpassQuery = OverpassQueryCreator(oqlElementNames, tagExprRoot).create()
    }

    /** returns whether the given element is found through (=matches) this expression */
    fun matches(element: Element) =
        when (element.type) {
            Element.Type.NODE -> elementsTypes.contains(NODES)
            Element.Type.WAY -> elementsTypes.contains(WAYS)
            Element.Type.RELATION -> elementsTypes.contains(RELATIONS)
            else -> false
        } && (tagExprRoot?.matches(element.tags) ?: true)

    /** returns this expression as a Overpass query string */
    fun toOverpassQLString(): String {
        return overpassQuery
    }

}
